src/GraphControl/Control/Grapher.cpp

```diff
@@ -389,4 +434,98 @@ namespace GraphControl
+        }
+
+        // For scaling, the graphing engine interprets x,y position between the range [-1, 1].
+        // Translate the pointer position to the [-1, 1] bounds.
+        const auto& pos = currentPointer->Position;
+        const auto[centerX, centerY] = PointerPositionToGraphPosition(pos.X, pos.Y, ActualWidth, ActualHeight);
+
+        ScaleRange(centerX, centerY, scale);
+
+        args->Handled = true;
+    }
+
+    void Grapher::OnPointerPressed(PointerRoutedEventArgs^ e)
+	{
+        // Set the pointer capture to the element being interacted with so that only it
+        // will fire pointer-related events
+        CapturePointer(e->Pointer);
+	}
+
+    void Grapher::OnPointerReleased(PointerRoutedEventArgs^ e)
+	{
+        // Release the pointer
+        ReleasePointerCapture(e->Pointer);
+	}
+
+    void Grapher::OnPointerCanceled(PointerRoutedEventArgs^ e)
+    {
+        ReleasePointerCapture(e->Pointer);
+    }
+
+    void Grapher::OnManipulationDelta(ManipulationDeltaRoutedEventArgs^ args)
+    {
+        if (m_graph)
+        {
+            if (auto renderer = m_graph->GetRenderer())
+            {
+                const double width = ActualWidth;
+                const double height = ActualHeight;
+
+                const auto& translation = args->Delta.Translation;
+                double translationX = translation.X;
+                double translationY = translation.Y;
+                if (translationX != 0 || translation.Y != 0)
+                {
+                    // The graphing engine pans the graph according to a ratio for x and y.
+                    // A value of +1 means move a half screen in the positive direction for the given axis.
+                    // Convert the manipulation's translation values to ratios for the engine.
+                    translationX /= -width;
+                    translationY /= height;
+
+                    if (FAILED(renderer->MoveRangeByRatio(translationX, translationY)))
+                    {
+                        return;
+                    }
+                }
+
+                if (double scale = args->Delta.Scale; scale != 1.0)
+                {
+                    // The graphing engine interprets scale amounts as the inverse of the value retrieved
+                    // from the ManipulationUpdatedEventArgs. Invert the scale amount for the engine.
+                    scale = 1.0 / scale;
+
+                    // The graphing engine interprets x,y position between the range [-1, 1].
+                    // Translate the pointer position to the [-1, 1] bounds.
+                    const auto& pos = args->Position;
+                    const auto[centerX, centerY] = PointerPositionToGraphPosition(pos.X, pos.Y, width, height);
+
+                    if (FAILED(renderer->ScaleRange(centerX, centerY, scale)))
+                    {
+                        return;
+                    }
+                }
+
+                m_renderMain->RunRenderPass();
```

Comment:
Is it possible for neither PointerPositionToGraphPosition and MoveRangeByRatio to not be called? If so do we still need to call RunRenderPass?

Reply:
The reason is that those public-facing functions call a render pass before they return.  If I used them here, that would mean a single manipulation event would result in two render passes which is suboptimal.  Currently there aren't any clients of the public-facing functions so I can just remove them.  I used them when I was prototyping and playing around with the rendering APIs.

Reply:
I'm not sure this is possible because _something_ happened to generate a ManipulationUpdated event.  I added the optimization just in case.